File: dsa/examples/use-cases/firmware-signing/sign_firmware.py
#!/usr/bin/env python3
"""
Post-Quantum Firmware Signing Tool

Sign firmware images using post-quantum digital signatures (ML-DSA or SLH-DSA)
for secure boot and OTA update verification.

Features:
- Firmware version tracking for rollback protection
- Device/hardware compatibility metadata
- Manifest generation for OTA updates
- Support for multiple signature algorithms

Usage:
    python sign_firmware.py --key secret.key --firmware firmware.bin [options]

Output:
    Creates firmware.bin.fwsig (signed firmware manifest)
"""

import argparse
import hashlib
import json
import sys
import struct
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional
from dataclasses import dataclass, asdict

# Try to import from installed package first, fall back to source path
try:
    from mldsa import MLDSA44, MLDSA65, MLDSA87
    from slhdsa import (
        SLHDSA_SHAKE_128f, SLHDSA_SHAKE_128s,
        SLHDSA_SHAKE_192f, SLHDSA_SHAKE_256f,
    )
except ImportError:
    sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent / "src" / "py"))
    from mldsa import MLDSA44, MLDSA65, MLDSA87
    from slhdsa import (
        SLHDSA_SHAKE_128f, SLHDSA_SHAKE_128s,
        SLHDSA_SHAKE_192f, SLHDSA_SHAKE_256f,
    )


# Supported algorithms
ALGORITHMS = {
    "mldsa44": {"class": MLDSA44, "name": "ML-DSA-44", "security": "NIST Level 2"},
    "mldsa65": {"class": MLDSA65, "name": "ML-DSA-65", "security": "NIST Level 3"},
    "mldsa87": {"class": MLDSA87, "name": "ML-DSA-87", "security": "NIST Level 5"},
    "slh-shake-128f": {"class": SLHDSA_SHAKE_128f, "name": "SLH-DSA-SHAKE-128f", "security": "NIST Level 1"},
    "slh-shake-128s": {"class": SLHDSA_SHAKE_128s, "name": "SLH-DSA-SHAKE-128s", "security": "NIST Level 1"},
    "slh-shake-192f": {"class": SLHDSA_SHAKE_192f, "name": "SLH-DSA-SHAKE-192f", "security": "NIST Level 3"},
    "slh-shake-256f": {"class": SLHDSA_SHAKE_256f, "name": "SLH-DSA-SHAKE-256f", "security": "NIST Level 5"},
}


@dataclass
class FirmwareMetadata:
    """Firmware metadata for signing."""
    version: str
    version_code: int  # Numeric version for rollback protection
    device_type: str
    hardware_rev: str
    build_date: str
    build_id: str
    description: str = ""
    min_bootloader_version: str = ""
    compatibility: list = None  # List of compatible device models

    def __post_init__(self):
        if self.compatibility is None:
            self.compatibility = []


def compute_firmware_hash(filepath: Path) -> dict:
    """
    Compute multiple hashes of firmware image.

    Returns SHA-256 and SHA-512 for redundancy.
    """
    sha256 = hashlib.sha256()
    sha512 = hashlib.sha512()

    with open(filepath, "rb") as f:
        for chunk in iter(lambda: f.read(65536), b""):
            sha256.update(chunk)
            sha512.update(chunk)

    return {
        "sha256": sha256.hexdigest(),
        "sha512": sha512.hexdigest(),
    }


def parse_version(version_str: str) -> int:
    """
    Convert version string to numeric code for rollback protection.

    Supports formats: "1.2.3", "1.2", "1"
    """
    parts = version_str.split(".")
    code = 0
    multipliers = [1000000, 1000, 1]  # Major.Minor.Patch

    for i, part in enumerate(parts[:3]):
        try:
            code += int(part) * multipliers[i]
        except ValueError:
            pass

    return code


def detect_algorithm(key_bytes: bytes) -> str:
    """Detect algorithm from secret key size."""
    key_sizes = {
        2560: "mldsa44",
        4032: "mldsa65",
        4896: "mldsa87",
    }

    size = len(key_bytes)
    if size in key_sizes:
        return key_sizes[size]

    if 64 <= size <= 128:
        return "slh-shake-128f"

    raise ValueError(f"Cannot detect algorithm from key size {size} bytes")


def load_secret_key(key_path: Path, password: Optional[str] = None) -> tuple[bytes, dict]:
    """Load secret key and certificate metadata."""
    key_data = key_path.read_bytes()

    if key_data.startswith(b'{'):
        key_json = json.loads(key_data)
        if "encrypted" in key_json and key_json["encrypted"]:
            if not password:
                raise ValueError("Key is encrypted but no password provided")
            raise NotImplementedError("Encrypted key support requires password decryption")
        key_data = bytes.fromhex(key_json.get("key", key_json.get("secret_key", "")))

    cert_path = key_path.parent / key_path.name.replace("_secret.key", "_certificate.json")
    cert_data = {}
    if cert_path.exists():
        cert_data = json.loads(cert_path.read_text())

    return key_data, cert_data


def sign_firmware(
    filepath: Path,
    secret_key: bytes,
    algorithm: str,
    metadata: FirmwareMetadata,
    signer_info: dict,
    context: bytes = b"firmware",
) -> dict:
    """
    Sign firmware image and generate signed manifest.

    Args:
        filepath: Path to firmware binary
        secret_key: Secret signing key
        algorithm: Algorithm identifier
        metadata: Firmware metadata
        signer_info: Signer identity
        context: Context string for domain separation

    Returns:
        Signed firmware manifest dictionary
    """
    if algorithm not in ALGORITHMS:
        raise ValueError(f"Unknown algorithm: {algorithm}")

    algo_info = ALGORITHMS[algorithm]

    # Compute firmware hashes
    hashes = compute_firmware_hash(filepath)
    file_size = filepath.stat().st_size

    # Generate timestamp
    timestamp = datetime.now(timezone.utc).isoformat()

    # Create the message to sign (canonical JSON)
    sign_data = {
        "firmware": {
            "name": filepath.name,
            "size": file_size,
            "hashes": hashes,
        },
        "metadata": asdict(metadata),
        "timestamp": timestamp,
    }
    sign_message = json.dumps(sign_data, sort_keys=True).encode("utf-8")

    # Sign the message
    signer = algo_info["class"]()
    signature = signer.sign(secret_key, sign_message, ctx=context)

    # Build the complete manifest
    manifest = {
        "manifest_version": "1.0",
        "type": "firmware-signature",
        "algorithm": {
            "id": algorithm,
            "name": algo_info["name"],
            "security_level": algo_info["security"],
            "standard": "FIPS 204" if algorithm.startswith("mldsa") else "FIPS 205",
        },
        "firmware": {
            "name": filepath.name,
            "size": file_size,
            "hashes": hashes,
        },
        "metadata": asdict(metadata),
        "signature": {
            "value": signature.hex(),
            "encoding": "hex",
            "context": context.hex(),
        },
        "timestamp": timestamp,
        "signer": signer_info,
        "security": {
            "rollback_protection": True,
            "minimum_version_code": metadata.version_code,
        },
    }

    return manifest


def main():
    parser = argparse.ArgumentParser(
        description="Sign firmware images with post-quantum signatures",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Sign firmware with version info
  python sign_firmware.py --key signing.key --firmware firmware.bin \\
      --version 2.1.0 --device-type "IoT-Sensor-v1"

  # Sign with full metadata
  python sign_firmware.py --key signing.key --firmware update.bin \\
      --version 3.0.0 --device-type "Gateway" --hardware-rev "rev-c" \\
      --build-id "build-12345" --description "Security update"

  # Sign for specific hardware compatibility
  python sign_firmware.py --key signing.key --firmware firmware.bin \\
      --version 1.5.0 --device-type "Sensor" \\
      --compatible "Model-A" --compatible "Model-B"
        """,
    )

    parser.add_argument(
        "--key", "-k",
        required=True,
        type=Path,
        help="Path to secret key file",
    )
    parser.add_argument(
        "--firmware", "-f",
        required=True,
        type=Path,
        help="Path to firmware binary",
    )
    parser.add_argument(
        "--algorithm", "-a",
        choices=list(ALGORITHMS.keys()),
        help="Signing algorithm (auto-detected from key if not specified)",
    )
    parser.add_argument(
        "--output", "-o",
        type=Path,
        help="Output manifest file (default: <firmware>.fwsig)",
    )

    # Firmware metadata
    parser.add_argument(
        "--version", "-v",
        required=True,
        help="Firmware version (e.g., 2.1.0)",
    )
    parser.add_argument(
        "--device-type",
        required=True,
        help="Device type identifier",
    )
    parser.add_argument(
        "--hardware-rev",
        default="",
        help="Hardware revision (e.g., rev-c)",
    )
    parser.add_argument(
        "--build-id",
        default="",
        help="Build identifier",
    )
    parser.add_argument(
        "--build-date",
        help="Build date (ISO format, defaults to now)",
    )
    parser.add_argument(
        "--description",
        default="",
        help="Firmware description",
    )
    parser.add_argument(
        "--min-bootloader",
        default="",
        help="Minimum bootloader version required",
    )
    parser.add_argument(
        "--compatible",
        action="append",
        default=[],
        help="Compatible device model (can specify multiple)",
    )

    # Signer info
    parser.add_argument(
        "--signer-name",
        help="Signer name",
    )
    parser.add_argument(
        "--signer-org",
        help="Signer organization",
    )

    parser.add_argument(
        "--password", "-p",
        help="Password for encrypted key",
    )
    parser.add_argument(
        "--quiet", "-q",
        action="store_true",
        help="Suppress output except errors",
    )

    args = parser.parse_args()

    # Validate inputs
    if not args.key.exists():
        print(f"Error: Key file not found: {args.key}", file=sys.stderr)
        sys.exit(1)

    if not args.firmware.exists():
        print(f"Error: Firmware file not found: {args.firmware}", file=sys.stderr)
        sys.exit(1)

    # Load key
    try:
        secret_key, cert_data = load_secret_key(args.key, args.password)
    except Exception as e:
        print(f"Error loading key: {e}", file=sys.stderr)
        sys.exit(1)

    # Determine algorithm
    algorithm = args.algorithm
    if not algorithm:
        try:
            algorithm = detect_algorithm(secret_key)
            if not args.quiet:
                print(f"Auto-detected algorithm: {algorithm}")
        except ValueError as e:
            print(f"Error: {e}. Please specify --algorithm", file=sys.stderr)
            sys.exit(1)

    # Build metadata
    build_date = args.build_date or datetime.now(timezone.utc).strftime("%Y-%m-%d")
    metadata = FirmwareMetadata(
        version=args.version,
        version_code=parse_version(args.version),
        device_type=args.device_type,
        hardware_rev=args.hardware_rev,
        build_date=build_date,
        build_id=args.build_id or f"build-{datetime.now().strftime('%Y%m%d%H%M%S')}",
        description=args.description,
        min_bootloader_version=args.min_bootloader,
        compatibility=args.compatible,
    )

    # Build signer info
    signer_info = {}
    if cert_data:
        subject = cert_data.get("subject", {})
        signer_info = {
            "name": args.signer_name or subject.get("common_name", ""),
            "organization": args.signer_org or subject.get("organization", ""),
        }
    else:
        signer_info = {
            "name": args.signer_name or "Unknown",
            "organization": args.signer_org or "",
        }
    signer_info = {k: v for k, v in signer_info.items() if v}

    # Sign the firmware
    try:
        manifest = sign_firmware(
            filepath=args.firmware,
            secret_key=secret_key,
            algorithm=algorithm,
            metadata=metadata,
            signer_info=signer_info,
        )
    except Exception as e:
        print(f"Error signing firmware: {e}", file=sys.stderr)
        sys.exit(1)

    # Write manifest
    output_path = args.output or args.firmware.with_suffix(args.firmware.suffix + ".fwsig")
    try:
        with open(output_path, "w") as f:
            json.dump(manifest, f, indent=2)
    except Exception as e:
        print(f"Error writing manifest: {e}", file=sys.stderr)
        sys.exit(1)

    if not args.quiet:
        print(f"\nFirmware signed successfully!")
        print(f"  Firmware:     {args.firmware}")
        print(f"  Version:      {metadata.version} (code: {metadata.version_code})")
        print(f"  Device Type:  {metadata.device_type}")
        print(f"  SHA-256:      {manifest['firmware']['hashes']['sha256'][:16]}...")
        print(f"  Algorithm:    {manifest['algorithm']['name']}")
        print(f"  Output:       {output_path}")


if __name__ == "__main__":
    main()

File: dsa/examples/use-cases/firmware-signing/sign_firmware.cpp
/**
 * Post-Quantum Firmware Signing Tool (C++)
 *
 * Signs firmware images for secure boot and OTA updates, producing a
 * JSON manifest with rollback protection and device compatibility info.
 *
 * Usage:
 *   ./sign_firmware <secret_key> <firmware> --version <ver> --device-type <type> [options]
 *
 * Options:
 *   --version, -v <ver>      Firmware version (e.g., 2.1.0) [required]
 *   --device-type <type>     Device type identifier [required]
 *   --algorithm <alg>        Algorithm (auto-detected from key if omitted)
 *   --output <file>          Output manifest file (default: <firmware>.fwsig)
 *   --hardware-rev <rev>     Hardware revision
 *   --build-id <id>          Build identifier
 *   --description <desc>     Firmware description
 *   --min-bootloader <ver>   Minimum bootloader version
 *   --compatible <model>     Compatible device model (repeatable)
 *   --signer-name <name>     Signer name
 *   --signer-org <org>       Signer organization
 *   --quiet                  Suppress output
 *
 * Example:
 *   ./sign_firmware keys/fw_secret.key firmware.bin -v 2.1.0 --device-type "IoT-Sensor"
 */

#include <iostream>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <chrono>
#include <vector>
#include <string>
#include <algorithm>
#include <stdexcept>

#include "mldsa/mldsa.hpp"
#include "slhdsa/slh_dsa.hpp"

#include <openssl/sha.h>

// Read binary file
std::vector<uint8_t> read_file(const std::string& path) {
    std::ifstream file(path, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Cannot open file: " + path);
    }
    return std::vector<uint8_t>(
        std::istreambuf_iterator<char>(file),
        std::istreambuf_iterator<char>()
    );
}

// Write text file
bool write_text_file(const std::string& path, const std::string& content) {
    std::ofstream file(path);
    if (!file) return false;
    file << content;
    return file.good();
}

// Convert bytes to hex string
std::string to_hex(const std::vector<uint8_t>& data) {
    std::ostringstream oss;
    for (uint8_t b : data) {
        oss << std::hex << std::setfill('0') << std::setw(2) << static_cast<int>(b);
    }
    return oss.str();
}

// Compute SHA-256 hash
std::vector<uint8_t> sha256(const std::vector<uint8_t>& data) {
    std::vector<uint8_t> hash(SHA256_DIGEST_LENGTH);
    SHA256(data.data(), data.size(), hash.data());
    return hash;
}

// Compute SHA-512 hash
std::vector<uint8_t> sha512(const std::vector<uint8_t>& data) {
    std::vector<uint8_t> hash(SHA512_DIGEST_LENGTH);
    SHA512(data.data(), data.size(), hash.data());
    return hash;
}

// Get current ISO timestamp
std::string get_timestamp() {
    auto now = std::chrono::system_clock::now();
    auto time = std::chrono::system_clock::to_time_t(now);
    std::stringstream ss;
    ss << std::put_time(std::gmtime(&time), "%Y-%m-%dT%H:%M:%S+00:00");
    return ss.str();
}

// Extract filename from path
std::string get_filename(const std::string& path) {
    size_t pos = path.find_last_of("/\\");
    return (pos == std::string::npos) ? path : path.substr(pos + 1);
}

// JSON escape string
std::string json_escape(const std::string& s) {
    std::string result;
    for (char c : s) {
        switch (c) {
            case '"': result += "\\\""; break;
            case '\\': result += "\\\\"; break;
            case '\n': result += "\\n"; break;
            case '\r': result += "\\r"; break;
            case '\t': result += "\\t"; break;
            default: result += c;
        }
    }
    return result;
}

// Parse version string to numeric code (MAJOR*1000000 + MINOR*1000 + PATCH)
int parse_version(const std::string& version) {
    int major = 0, minor = 0, patch = 0;
    size_t start = 0;
    size_t dot1 = version.find('.');

    if (dot1 != std::string::npos) {
        major = std::stoi(version.substr(0, dot1));
        size_t dot2 = version.find('.', dot1 + 1);
        if (dot2 != std::string::npos) {
            minor = std::stoi(version.substr(dot1 + 1, dot2 - dot1 - 1));
            // Parse patch, ignoring any suffix like -beta
            std::string patch_str = version.substr(dot2 + 1);
            size_t suffix = patch_str.find_first_not_of("0123456789");
            if (suffix != std::string::npos) {
                patch_str = patch_str.substr(0, suffix);
            }
            if (!patch_str.empty()) {
                patch = std::stoi(patch_str);
            }
        } else {
            minor = std::stoi(version.substr(dot1 + 1));
        }
    } else {
        major = std::stoi(version);
    }

    return major * 1000000 + minor * 1000 + patch;
}

// Algorithm info structure
struct AlgorithmInfo {
    std::string id;
    std::string name;
    std::string security_level;
    std::string standard;
};

// Get algorithm info from name
AlgorithmInfo get_algorithm_info(const std::string& alg) {
    std::string lower = alg;
    std::transform(lower.begin(), lower.end(), lower.begin(), ::tolower);

    if (lower == "mldsa44") {
        return {"mldsa44", "ML-DSA-44", "NIST Level 2", "FIPS 204"};
    } else if (lower == "mldsa65") {
        return {"mldsa65", "ML-DSA-65", "NIST Level 3", "FIPS 204"};
    } else if (lower == "mldsa87") {
        return {"mldsa87", "ML-DSA-87", "NIST Level 5", "FIPS 204"};
    } else if (lower == "slh-shake-128f") {
        return {"slh-shake-128f", "SLH-DSA-SHAKE-128f", "NIST Level 1", "FIPS 205"};
    } else if (lower == "slh-shake-192f") {
        return {"slh-shake-192f", "SLH-DSA-SHAKE-192f", "NIST Level 3", "FIPS 205"};
    } else if (lower == "slh-shake-256f") {
        return {"slh-shake-256f", "SLH-DSA-SHAKE-256f", "NIST Level 5", "FIPS 205"};
    }
    throw std::invalid_argument("Unknown algorithm: " + alg);
}

// Detect algorithm from secret key size
std::string detect_algorithm(size_t sk_size) {
    switch (sk_size) {
        case 2560: return "mldsa44";
        case 4032: return "mldsa65";
        case 4896: return "mldsa87";
        case 64: return "slh-shake-128f";
        case 96: return "slh-shake-192f";
        case 128: return "slh-shake-256f";
        default:
            throw std::invalid_argument("Cannot detect algorithm from key size: " +
                                        std::to_string(sk_size));
    }
}

// Sign function template for ML-DSA
template<typename DSA>
std::vector<uint8_t> sign_mldsa(const std::vector<uint8_t>& sk,
                                 const std::vector<uint8_t>& message,
                                 const std::vector<uint8_t>& ctx) {
    DSA dsa;
    return dsa.sign(sk, message, ctx);
}

// Sign function template for SLH-DSA
template<typename DSA>
std::vector<uint8_t> sign_slhdsa(const std::vector<uint8_t>& sk,
                                  const std::vector<uint8_t>& message,
                                  const std::vector<uint8_t>& ctx) {
    DSA dsa;
    return dsa.sign(sk, message, ctx);
}

// Firmware metadata structure
struct FirmwareMetadata {
    std::string version;
    int version_code = 0;
    std::string device_type;
    std::string hardware_rev;
    std::string build_id;
    std::string description;
    std::string min_bootloader;
    std::vector<std::string> compatible;
    std::string signer_name;
    std::string signer_org;
};

// Create firmware manifest JSON
std::string create_manifest(const std::string& filename,
                           size_t file_size,
                           const std::string& sha256_hash,
                           const std::string& sha512_hash,
                           const std::string& timestamp,
                           const AlgorithmInfo& alg,
                           const std::vector<uint8_t>& signature,
                           const FirmwareMetadata& metadata) {
    std::ostringstream json;
    json << "{\n";
    json << "  \"manifest_version\": \"1.0\",\n";
    json << "  \"type\": \"firmware-signature\",\n";
    json << "  \"algorithm\": {\n";
    json << "    \"id\": \"" << alg.id << "\",\n";
    json << "    \"name\": \"" << alg.name << "\",\n";
    json << "    \"security_level\": \"" << alg.security_level << "\",\n";
    json << "    \"standard\": \"" << alg.standard << "\"\n";
    json << "  },\n";
    json << "  \"firmware\": {\n";
    json << "    \"name\": \"" << json_escape(filename) << "\",\n";
    json << "    \"size\": " << file_size << ",\n";
    json << "    \"hashes\": {\n";
    json << "      \"sha256\": \"" << sha256_hash << "\",\n";
    json << "      \"sha512\": \"" << sha512_hash << "\"\n";
    json << "    }\n";
    json << "  },\n";
    json << "  \"metadata\": {\n";
    json << "    \"version\": \"" << json_escape(metadata.version) << "\",\n";
    json << "    \"version_code\": " << metadata.version_code << ",\n";
    json << "    \"device_type\": \"" << json_escape(metadata.device_type) << "\"";

    if (!metadata.hardware_rev.empty()) {
        json << ",\n    \"hardware_rev\": \"" << json_escape(metadata.hardware_rev) << "\"";
    }
    if (!metadata.build_id.empty()) {
        json << ",\n    \"build_id\": \"" << json_escape(metadata.build_id) << "\"";
    }
    if (!metadata.description.empty()) {
        json << ",\n    \"description\": \"" << json_escape(metadata.description) << "\"";
    }
    if (!metadata.min_bootloader.empty()) {
        json << ",\n    \"min_bootloader_version\": \"" << json_escape(metadata.min_bootloader) << "\"";
    }
    if (!metadata.compatible.empty()) {
        json << ",\n    \"compatibility\": [";
        for (size_t i = 0; i < metadata.compatible.size(); ++i) {
            if (i > 0) json << ", ";
            json << "\"" << json_escape(metadata.compatible[i]) << "\"";
        }
        json << "]";
    }

    json << "\n  },\n";
    json << "  \"signature\": {\n";
    json << "    \"value\": \"" << to_hex(signature) << "\",\n";
    json << "    \"encoding\": \"hex\",\n";
    json << "    \"context\": \"" << to_hex(std::vector<uint8_t>{'f','i','r','m','w','a','r','e'}) << "\"\n";
    json << "  },\n";
    json << "  \"timestamp\": \"" << timestamp << "\"";

    if (!metadata.signer_name.empty() || !metadata.signer_org.empty()) {
        json << ",\n  \"signer\": {\n";
        if (!metadata.signer_name.empty()) {
            json << "    \"name\": \"" << json_escape(metadata.signer_name) << "\"";
            if (!metadata.signer_org.empty()) json << ",\n";
            else json << "\n";
        }
        if (!metadata.signer_org.empty()) {
            json << "    \"organization\": \"" << json_escape(metadata.signer_org) << "\"\n";
        }
        json << "  }";
    }

    json << ",\n  \"security\": {\n";
    json << "    \"rollback_protection\": true,\n";
    json << "    \"minimum_version_code\": " << metadata.version_code << "\n";
    json << "  }";

    json << "\n}\n";
    return json.str();
}

void print_usage() {
    std::cout << "Post-Quantum Firmware Signing Tool (C++)\n";
    std::cout << std::string(60, '=') << "\n\n";
    std::cout << "Usage: sign_firmware <secret_key> <firmware> -v <version> --device-type <type> [options]\n\n";
    std::cout << "Required:\n";
    std::cout << "  <secret_key>          Secret key file\n";
    std::cout << "  <firmware>            Firmware binary to sign\n";
    std::cout << "  -v, --version <ver>   Firmware version (e.g., 2.1.0)\n";
    std::cout << "  --device-type <type>  Device type identifier\n\n";
    std::cout << "Options:\n";
    std::cout << "  --algorithm <alg>        Algorithm (auto-detected from key)\n";
    std::cout << "  --output <file>          Output manifest (default: <firmware>.fwsig)\n";
    std::cout << "  --hardware-rev <rev>     Hardware revision\n";
    std::cout << "  --build-id <id>          Build identifier\n";
    std::cout << "  --description <desc>     Firmware description\n";
    std::cout << "  --min-bootloader <ver>   Minimum bootloader version\n";
    std::cout << "  --compatible <model>     Compatible device model (repeatable)\n";
    std::cout << "  --signer-name <name>     Signer name\n";
    std::cout << "  --signer-org <org>       Signer organization\n";
    std::cout << "  --quiet                  Suppress output\n";
    std::cout << "  --help                   Show this help\n\n";
    std::cout << "Algorithms:\n";
    std::cout << "  mldsa44, mldsa65, mldsa87 (ML-DSA, FIPS 204)\n";
    std::cout << "  slh-shake-128f, slh-shake-192f, slh-shake-256f (SLH-DSA, FIPS 205)\n\n";
    std::cout << "Example:\n";
    std::cout << "  ./sign_firmware keys/fw_secret.key firmware.bin -v 2.1.0 --device-type \"IoT-Sensor\"\n";
}

int main(int argc, char* argv[]) {
    if (argc < 5) {
        print_usage();
        return (argc == 1) ? 0 : 1;
    }

    std::string sk_file;
    std::string fw_file;
    std::string algorithm;
    std::string output_file;
    FirmwareMetadata metadata;
    bool quiet = false;

    // Parse arguments
    int i = 1;
    while (i < argc) {
        std::string arg = argv[i];

        if ((arg == "-v" || arg == "--version") && i + 1 < argc) {
            metadata.version = argv[++i];
        } else if (arg == "--device-type" && i + 1 < argc) {
            metadata.device_type = argv[++i];
        } else if (arg == "--algorithm" && i + 1 < argc) {
            algorithm = argv[++i];
        } else if (arg == "--output" && i + 1 < argc) {
            output_file = argv[++i];
        } else if (arg == "--hardware-rev" && i + 1 < argc) {
            metadata.hardware_rev = argv[++i];
        } else if (arg == "--build-id" && i + 1 < argc) {
            metadata.build_id = argv[++i];
        } else if (arg == "--description" && i + 1 < argc) {
            metadata.description = argv[++i];
        } else if (arg == "--min-bootloader" && i + 1 < argc) {
            metadata.min_bootloader = argv[++i];
        } else if (arg == "--compatible" && i + 1 < argc) {
            metadata.compatible.push_back(argv[++i]);
        } else if (arg == "--signer-name" && i + 1 < argc) {
            metadata.signer_name = argv[++i];
        } else if (arg == "--signer-org" && i + 1 < argc) {
            metadata.signer_org = argv[++i];
        } else if (arg == "--quiet" || arg == "-q") {
            quiet = true;
        } else if (arg == "--help" || arg == "-h") {
            print_usage();
            return 0;
        } else if (arg[0] != '-') {
            if (sk_file.empty()) {
                sk_file = arg;
            } else if (fw_file.empty()) {
                fw_file = arg;
            }
        } else {
            std::cerr << "Unknown option: " << arg << "\n";
            return 1;
        }
        ++i;
    }

    // Validate required arguments
    if (sk_file.empty()) {
        std::cerr << "Error: Secret key is required\n";
        return 1;
    }
    if (fw_file.empty()) {
        std::cerr << "Error: Firmware file is required\n";
        return 1;
    }
    if (metadata.version.empty()) {
        std::cerr << "Error: Version is required (-v/--version)\n";
        return 1;
    }
    if (metadata.device_type.empty()) {
        std::cerr << "Error: Device type is required (--device-type)\n";
        return 1;
    }

    try {
        // Parse version to code
        metadata.version_code = parse_version(metadata.version);

        // Read secret key
        if (!quiet) std::cerr << "Reading secret key: " << sk_file << "\n";
        auto sk = read_file(sk_file);

        // Auto-detect algorithm if not specified
        if (algorithm.empty()) {
            algorithm = detect_algorithm(sk.size());
            if (!quiet) std::cerr << "Detected algorithm: " << algorithm << "\n";
        }

        // Read firmware
        if (!quiet) std::cerr << "Reading firmware: " << fw_file << "\n";
        auto fw_data = read_file(fw_file);

        // Compute hashes
        auto hash256 = sha256(fw_data);
        auto hash512 = sha512(fw_data);
        std::string hash256_hex = to_hex(hash256);
        std::string hash512_hex = to_hex(hash512);

        // Get timestamp
        std::string timestamp = get_timestamp();

        // Build signed message
        std::string filename = get_filename(fw_file);
        std::ostringstream msg_stream;
        msg_stream << "{\"firmware_hash\":\"" << hash256_hex << "\",";
        msg_stream << "\"firmware_name\":\"" << json_escape(filename) << "\",";
        msg_stream << "\"firmware_size\":" << fw_data.size() << ",";
        msg_stream << "\"version\":\"" << json_escape(metadata.version) << "\",";
        msg_stream << "\"version_code\":" << metadata.version_code << ",";
        msg_stream << "\"device_type\":\"" << json_escape(metadata.device_type) << "\",";
        msg_stream << "\"timestamp\":\"" << timestamp << "\"}";
        std::string msg_str = msg_stream.str();
        std::vector<uint8_t> message(msg_str.begin(), msg_str.end());

        // Firmware context for domain separation
        std::vector<uint8_t> ctx = {'f', 'i', 'r', 'm', 'w', 'a', 'r', 'e'};

        // Sign
        if (!quiet) std::cerr << "Signing with " << algorithm << "...\n";
        std::vector<uint8_t> signature;
        AlgorithmInfo alg_info = get_algorithm_info(algorithm);

        if (algorithm == "mldsa44") {
            signature = sign_mldsa<mldsa::MLDSA44>(sk, message, ctx);
        } else if (algorithm == "mldsa65") {
            signature = sign_mldsa<mldsa::MLDSA65>(sk, message, ctx);
        } else if (algorithm == "mldsa87") {
            signature = sign_mldsa<mldsa::MLDSA87>(sk, message, ctx);
        } else if (algorithm == "slh-shake-128f") {
            signature = sign_slhdsa<slhdsa::SLHDSA_SHAKE_128f>(sk, message, ctx);
        } else if (algorithm == "slh-shake-192f") {
            signature = sign_slhdsa<slhdsa::SLHDSA_SHAKE_192f>(sk, message, ctx);
        } else if (algorithm == "slh-shake-256f") {
            signature = sign_slhdsa<slhdsa::SLHDSA_SHAKE_256f>(sk, message, ctx);
        } else {
            throw std::invalid_argument("Unsupported algorithm: " + algorithm);
        }

        if (!quiet) std::cerr << "Signature size: " << signature.size() << " bytes\n";

        // Create manifest
        std::string manifest = create_manifest(
            filename, fw_data.size(), hash256_hex, hash512_hex, timestamp,
            alg_info, signature, metadata
        );

        // Write output
        if (output_file.empty()) {
            output_file = fw_file + ".fwsig";
        }

        if (!write_text_file(output_file, manifest)) {
            std::cerr << "Error: Failed to write manifest file\n";
            return 1;
        }

        if (!quiet) {
            std::cerr << "Manifest written to: " << output_file << "\n";
            std::cerr << "\nFirmware Signing Summary:\n";
            std::cerr << "  Version:      " << metadata.version << " (code: " << metadata.version_code << ")\n";
            std::cerr << "  Device Type:  " << metadata.device_type << "\n";
            std::cerr << "  Size:         " << fw_data.size() << " bytes\n";
            std::cerr << "  Algorithm:    " << alg_info.name << "\n";
            std::cerr << "\nFirmware signed successfully!\n";
        }

        return 0;

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
}
